lua/powerups/server/phoenix.moon

```diff
@@ -0,0 +1,239 @@
+get: getConf = CFCPowerups.Config
+
+import Rand, cos, sin from math
+import Create from ents
+import SpriteTrail from util
+
+export PhoenixPowerup
+class PhoenixPowerup extends BasePowerup
+    @powerupID: "powerup_phoenix"
+
+    @powerupWeights:
+        tier1: 1
+        tier2: 1
+        tier3: 1
+        tier4: 1
+
+    new: (ply) =>
+        super ply
+
+        @UsesRemaining = getConf "phoenix_uses"
+        @reviveHealth = getConf "phoenix_revive_health"
+        @reviveArmor = getConf "phoenix_revive_armor"
+        @immunityDamageMult = getConf "phoenix_immunity_damage_multiplier"
+        @immunityDuration = getConf "phoenix_immunity_duration"
+
+        @setModel = (FindMetaTable "Entity").SetModel -- CFC blocks skeleton and charple by wrapping Player:SetModel()
+
+        @ApplyEffect!
+
+    Revive: =>
+        @UsesRemaining = @UsesRemaining - 1
+        @immune = true
+        @ownerModel = @owner\GetModel!
+
+        maxRegenHealth = math.min @reviveHealth, @owner\GetMaxHealth!
+        maxRegenArmor = math.min @reviveArmor, @owner\GetMaxArmor!
+
+        with @owner
+            \SetHealth 1
+            \SetArmor maxRegenArmor == 0 and 0 or 1
+            \ChatPrint "Like a phoenix, you rise from the ashes! (#{@UsesRemaining} uses remaining)"
+
+            splodePitch = math.random 80, 90
+            \EmitSound "ambient/levels/labs/electric_explosion4.wav", 75, splodePitch, 1
+            \ScreenFade SCREENFADE.IN, (Color 255, 240, 230, 150), 2, 0.1
+
+            util.ScreenShake \GetPos!, 10, 20, 2.5, 1500
+            util.ScreenShake \GetPos!, 40, 40, 0.5, 500
+
+        with eff = EffectData!
+            eff\SetOrigin @owner\GetPos!
+            eff\SetNormal Vector 0, 0, 1
+
+            util.Effect "VortDispel", eff, true, true
+            util.Effect "HL1GaussWallImpact2", eff, true, true
+
+        with eff = EffectData!
+            eff\SetOrigin @owner\WorldSpaceCenter!
+            eff\SetNormal Vector 0, 0, -1
+            eff\SetMagnitude 50
+
+            util.Effect "HL1GaussWallPunchExit", eff, true, true
+
+        with eff = EffectData!
+            eff\SetOrigin @owner\GetPos!
+            eff\SetMagnitude 1
+            eff\SetScale 1
+
+            util.Effect "cball_explode", eff, true, true
+
+        self.setModel @owner, "models/player/charple.mdl"
+
+        with @holyMusic = CreateSound @owner, "music/hl2_song10.mp3"
+            songPitch = math.random 140, 160
+            \PlayEx 1, songPitch
+            \FadeOut @immunityDuration
+
+        with @heartbeatSound = CreateSound @owner, "player/heartbeat1.wav"
+            \Play 75, 80
+
+        -- slowly regen health + armor from 0
+        timer.Create @regenTimerName, 0.1, 0, ->
+            bad = not @immune or not IsValid @owner
+
+            if bad then
+                timer.Remove @regenTimerName
+                return
```

Comment:
i think it reads a little better if you check `good` instead of `bad`
```suggestion
            good = @immune and IsValid @owner
            return timer.Remove @regenTimerName unless good
```